doc/doxygen/headers/glossary.h

```diff
@@ -1234,9 +1234,158 @@
+ *
+ * All of these examples use that the returned object is of a type whose
+ * set of possible values contains values that cannot be legitimate
+ * return values (in mathematical language: they are not part of the
+ * "range" of the function) and that can consequently be used to indicate
+ * errors. This is awkward because the mapping of error codes into the
+ * space of possible return values depends very much on the function and
+ * what it can and cannot return. For example, `sqrt()` could return `-1`
+ * as an error, but `sin()` can not because `-1` is a valid return value.
+ * Also, not all functions allow for this. For example, the
+ * [`strtol`](https://en.cppreference.com/w/c/string/byte/strtol)
+ * ("string to long (integer)") function takes a string as input and
+ * returns a long integer as output. But since clearly every possible
+ * value of the type `long int` can legitimately be returned, errors in
+ * the input (say, if someone provided the string `"nonsense"` as
+ * input) cannot be indicated via the return object, and the function
+ * needs to indicate errors through another mechanism. The C language
+ * does that by letting functions such as `strtol` set the global variable
+ * `errno` to a nonzero value to indicate an error (an approach that
+ * comes with its own set of problems, among which are that people
+ * tend to forget the value of this variable after calling the function).
+ *
+ * The examples listed above date back to the time when C was first
+ * developed, in the late 1960s and 1970s. C++ solves this conundrum
+ * in a more systematic way. First, functions can throw exceptions of
+ * any type, and one can think of a thrown exception as simply another
+ * possible return value of functions that indicates errors without
+ * requiring having a part of the value space of the return type of a
+ * function occupied for error codes. In fact, the type of an exception
+ * is completely decoupled from the usual return type: You can pass as
+ * much information through exceptions you throw, even if the function
+ * in question returns just a meager `int` in regular operation. This
+ * approach is used in a number of deal.II functions: If inputs don't
+ * make sense, the program is either aborted (typically via an
+ * `Assert` statement) if the inputs are believed to be hard-coded --
+ * say, when adding vectors of different length -- or an exception is
+ * thrown via C++'s `throw` statement. The function
+ * Mapping::transform_real_to_unit_cell() is an example of the latter.
+ * Second, in newer C++ standards, one can use the
+ * [std::expected<T,E>](https://en.cppreference.com/w/cpp/utility/expected)
+ * class as the return value that can be thought as "this function
+ * returns objects of type `T`, but if an error was detected, then the
+ * function instead returns an object of type `E`". You can then ask
+ * the returned object whether it stores one or the other.
```

Comment:
You could also use `std::optional` which would just tell you that a value was not provided.

Reply:
I think `optional` is a better choice here since it's either an integer or nothing - i.e., not an error type.

Reply:
`std::optional` is only an option in case you don't want to convey any reason for the failure. That's perhaps useful in some cases, but all of those functions that set `errno`, for example, do want to explain what goes wrong by setting that variable to a specific value. 

It is true that in some functions `std::optional` is the right choice, such as the `IndexSet` functions mentioned up above. I added this information to the glossary entry, explaining which one is useful when.